draft-peabody-dispatch-new-uuid-format-03.txt

```diff
@@ -323,1014 +349,540 @@ Internet-Draft               new-uuid-format               February 2022
-   with the available sub-second precision bits.
-
-   Examining Figure 3 one can observe:
-
-   *  The first 36 bits have been dedicated to the Unix Timestamp
-      (unixts)
-
-   *  All 12 bits of scenario subsec_a is fully dedicated to millisecond
-      information (msec).
-
-   *  The 4 Version bits remain unchanged (ver).
-
-   *  All 12 bits of subsec_b have been dedicated to a monotonic clock
-      sequence counter (seq).
-
-   *  The 2 Variant bits remain unchanged (var).
-
-   *  Finally the remaining 62 bits in the subsec_seq_node section are
-      layout is filled out with random data to pad the length and
-      provide guaranteed uniqueness (rand).
-
-        0                   1                   2                   3
-        0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-       |                            unixts                             |
-       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-       |unixts |         msec          |  ver  |          seq          |
-       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-       |var|                         rand                              |
-       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-       |                             rand                              |
-       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-
-   Figure 3: UUIDv7 Field and Bit Layout - Encoding Example (Millisecond
-                                 Precision)
-
-
-
-
-
-Peabody & Davis           Expires 5 August 2022                [Page 13]
-
-Internet-Draft               new-uuid-format               February 2022
-
-
-   In Figure 4 the UUIDv7 has been created with Microsecond precision
-   with the available sub-second precision bits.
-
-   Examining Figure 4 one can observe:
-
-   *  The first 36 bits have been dedicated to the Unix Timestamp
-      (unixts)
+                   Figure 2: UUIDv7 Field and Bit Layout
 
-   *  All 12 bits of scenario subsec_a is fully dedicated to providing
-      sub-second encoding for the Microsecond precision (usec).
+   unix_ts_ms:
+      48 bit big-endian unsigned number of milliseconds since midnight 1
+      January 1970 UTC.  Follows the same rules as a UNIX timestamp,
+      i.e. does not include leap seconds.
 
-   *  The 4 Version bits remain unchanged (ver).
+   rand_a:
+      16 bits pseudo-random data to provide uniqueness as per
+      Section 5.10
 
-   *  All 12 bits of subsec_b have been dedicated to providing sub-
-      second encoding for the Microsecond precision (usec).
+   var_ver:
+      The 8 bit combined variant and version field with the value 0xE7
+      as defined by Section 5.2.
 
-   *  The 2 Variant bits remain unchanged (var).
+   rand_b:
+      The final 56 bits of pseudo-random data to provide uniqueness as
+      per Section 5.10 and Section 5.7.
```

Comment:
Perhaps, it's rand_a that relates to Section 5.7, isn't it? The monotonic counter should be placed right after the timestamp so the generated IDs can be lexicographically sortable.

Just for clarification, do you have any intention (e.g. limiting counter size or ensuring minimum per-ID entropy) in allocating two rand_[ab] fields separately with the specific size for each? I was wondering if it is allowed to utilize a counter larger than 16 bits.

Reply:
> I'm okay to just get it over it and go with a milliseconds timestamp

At the risk of horsing a dead beat...

Is there a reason `unix_ts_ms` is 48 bits instead of 42 or 44?  44 bits would suffice for dates to year 2527.  I only mention this because the current verbiage ("big endian") means the first 6 high-order bits won't be used for nearly a century.  The first 4 for 500 years. 

Seems like a waste.  Maybe go with a 44 bit field, instead?
```
    0                   1                   2                   3
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                           unix_ts_ms                          |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |      unix_ts_ms       |                  rand_a               |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |    var_ver    |            rand_b                             |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                            rand_b                             |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

```

Reply:
> The monotonic counter should be placed right after the timestamp

After all the debate and discussion we've had about counters, would it be fair to say the consensus here is that counters are too problematic to be canonically addressed by this specification?   Summarizing why ...

* Counter uniqueness and monotonicity is limited in scope, typically to the host process.  But that scope is often smaller than the scope for which uniqueness is required. (Think many clients generating IDs, all connected to a central DB.)
* Counters have to accommodate worst case scenarios.  But such scenarios are relatively rare, resulting in fields that are largely static and fail to contribute to uniqueness as a result.  This is true of both statically and dynamically sized counter fields.
* Counter implementations tend to be complicated, requiring users to reason about UUID creation rates and counter rollover faults.

If this is the case, I'd suggest we just take a big fat "pass" on counters as far as the spec goes.  Don't even mention them in version 7.  And, for §5.7, limit ourselves to verbiage along these lines:

_"Sub-millisecond monotonicity may be provided at the discretion of implementors by utilizing the high-order bits of `rand_a` (and `rand_b`, if necessary) for an appropriate monotonic value.  For example, this may be a count of microseconds or nanoseconds, or of UUIDs generated, since the start of the current time interval._

However, such solutions should be approached with caution.  It is the consensus of the authors that such approaches are damaging to the _universal_ uniqueness guarantee this specification aspires to."_

Reply:
I am leaning towards including a clear guidance because counters need to be handled with caution.

BTW does a counter damage to the universal uniqueness? This is actually my long-time question. Assume a 16-bit field and two applications each of which generates 256 IDs in a millisecond. If the 16-bit field is filled with random bits, the probability of collision approximates to `1 - exp(-(512^2) / (2 * 2^16))` = 86.5%. On the other hand, if the 16-bit field is used as a randomly initialized counter, at least one collision occurs when the distance between two initial counter values is 256 or less, whose probability approximates to `2 * 256 / 2^16` = 0.8%. The downsides of the counter, though, are that burst collisions will occur at the 0.8% probability and that adjacent IDs are almost perfectly predictable.

My gut feeling is like: with a large enough counter, the universal collision probability decreases because fewer dice rolls are necessary. A similar but extreme example might be UUIDv1's pseudorandom node, which requires only one roll of dice for the life of a generator. But I am not yet able to solve this problem mathematically. Can anyone help with this?

Reply:
@LiosK I think you're comparing apples and oranges here.  That 86% # is the odds of a single collision in the case of 512 randomly selected values, while the 0.8% # is for a counter *spread* small enough to produce _at least_ one collision.  But small spreads will, on average, produce 128 collisions.  If you normalize the # of collisions by the probability you get an average of 0.86 collisions per 512 UUIDs in the random case, and 1.02 collisions per 512 UUIDs in the counter case.

Note, too, that you're not allowing for having to set the high counter bit to zero to mitigate the risk of rollover.  (But I believe you're also doubling the size of the spread needed to produce collisions, so I think your 0.8% number is overlapping spreads is actually correct. 😄 )

Reply:
Okay, math is too much work, but here's [a silly little script](https://codepen.io/broofa/pen/XWzEoxv?editors=0012) that  produces some statistically significant results (n = 10,000).  Covers cases for each of 2, 3, or 4 nodes generating 64, 256, and 1024 uuids.  

Seems like the difference is negligible for small #'s, but increases significantly (in favor of random values) as the number of uuids & nodes increases.


```
Uuids: 64, nNodes: 2
  Collisions/run (random): 0.13
  Collisions/run (counter): 0.14

Uuids: 256, nNodes: 2
  Collisions/run (random): 2.00
  Collisions/run (counter): 3.46

Uuids: 2048, nNodes: 2
  Collisions/run (random): 124.97
  Collisions/run (counter): 216.52

Uuids: 64, nNodes: 8
  Collisions/run (random): 1.98
  Collisions/run (counter): 6.76

Uuids: 256, nNodes: 8
  Collisions/run (random): 31.80
  Collisions/run (counter): 114.05

Uuids: 2048, nNodes: 8
  Collisions/run (random): 1888.09
  Collisions/run (counter): 6786.38

Uuids: 64, nNodes: 32
  Collisions/run (random): 31.72
  Collisions/run (counter): 125.08

Uuids: 256, nNodes: 32
  Collisions/run (random): 490.88
  Collisions/run (counter): 1941.97

Uuids: 2048, nNodes: 32
  Collisions/run (random): 24110.70
  Collisions/run (counter): 121532.89
```

Reply:
I found an interesting article about monotonic ULIDs. The author concluded that in his use case the monotonic ULIDs do not make collisions more likely. 

[How probable are collisions with ULID’s monotonic option?](https://medium.com/zendesk-engineering/how-probable-are-collisions-with-ulids-monotonic-option-d604d3ed2de)

Reply:
@fabiolimace Awesome! That's exactly the kind of article I have been looking for. Will take a look.

@broofa Small experiments don't necessarily reveal the truth in this case. I did some experiments before and they suggested that counters provided much better collision resistance than randomness when the counter field size is sufficiently larger than the number of concurrent nodes. It depends several factors including the number of nodes, the average number of IDs per node, and the distribution and concentration of such numbers across nodes. And we have to deal with statistically rare circumstances; that's why we do need math here.

Plus, we need to be clear which measure is more relevant to the universal uniqueness: the expected number of collisions or the probability of at least one collision. The latter seems to me more important for the use cases of UUID, but to reach a conclusion we need to evaluate how serious or problematic are the burst collisions that may occur by use of counters.

Reply:
> I found an interesting article about monotonic ULIDs.

Very interesting article. The math for when values are within a given range, _**k**_ , from one another is particularly insightful.  And I like how ULIDs treat the entire rand field as a monotonic counter.  'Makes a ton of sense.  However ... I believe the author has fallen afoul of the same mistake as @LiosK by not accounting for the # of collisions that occur when counters do end up overlapping.

For example, in the article's GO simulation ("Empirical Test" section) the `trial()` function exits immediately upon detecting a collision rather than counting how *many* collisions occur as IDs continue to be generated.  [Correcting for this](https://gist.github.com/broofa/d038e0ccbfa6f2c3cdcd415a17c37575) changes the behavior of the simulation significantly.

Running both the original version and my corrected version, I get the following results:

(Note: I ran w/ `TRIALS := 1000000` rather than `10000000` in both cases, so as to not have it take all day.  This increases the margin of error, but still shows the trends or lack thereof pretty clearly):

|Input | Original | Corrected |
|---|---|---|
| g=1 n=1000 k=0 p(n, k, d | 0.000127 | 0.000126 |
| g=1 n=1000 k=0 p(n, k, d | 0.000127 | 0.000126 |
| g=3 n=333 k=2 p(n, k, d | 7e-05 | 0.000106 |
| g=5 n=200 k=4 p(n, k, d | 5e-05 | 0.000133 |
| g=7 n=142 k=6 p(n, k, d | 3.7e-05 | 0.000079 |
| g=9 n=111 k=8 p(n, k, d | 3.3e-05 | 0.000146 |
| g=11 n=90 k=10 p(n, k, d | 1.6e-05 | 0.000109 |
| g=13 n=76 k=12 p(n, k, d | 1.8e-05 | 0.000117 |
| g=15 n=66 k=14 p(n, k, d | 1.8e-05 | 0.000135 |
| g=17 n=58 k=16 p(n, k, d | 1.3e-05 | 0.000132 |
| g=19 n=52 k=18 p(n, k, d | 1.4e-05 | 0.000104 |

  **I.e. Rather than the probability of collision decreasing as `k` increases, it stays more or less constant(!)**

Cc: @ggrossman You're the article author here, correct?  Would you care to comment?

> Small experiments don't necessarily reveal the truth in this case

Not necessarily, no.  But this isn't a valid reason for discounting such experiments, either.  And such experiments provide a concrete implementation that illustrates the assumptions authors may be making, as well as helping to identify trends that may be concerning as a system scales.  In that regard, I think they're both important and revealing.

For example, my "silly little script" experiment shows the trends in collisions between random vs. counter IDs as the number of nodes and uuids increases.  Is my model flawed?  Is there some reason to think these trends will reverse as the system scales?  Perhaps.  But if so, you and others have the code in front of you to throw darts at.

Reply:
Hi @broofa, neither the author or I made a mistake. We are just interested in the probability of no collision. We are trying to figure out if we can reasonably assume no collision will occur. From this perspective, 99.999% chance of zero collision vs. 99.999999% chance of zero collision is a huge difference, whereas we are somewhat indifferent to the number of collisions that would actually occur on the disaster day.

The probability of no collision and the expected number of collisions are both important properties when it comes to the universal uniqueness. My gut feeling is that the former is more relevant to UUID use cases, but I would hear what others say.

> In that regard, I think they're both important and revealing.

That's the point. I did some experiments too and found out that the problem wouldn't be solved without math. Neither experiments or math can solely reveal the truth. In this sense, we do need to discount experiments as well as math.

Reply:
@liosk: I vehemently disagree.

Your calculations aren't measuring id collision probability.  They're measuring id _group_ collision probability.

It's like me handing you a random ID generator and saying, "There's only a 0.0001% chance of collision."  But after using it for a few days, you discover that half your ids are duplicates.  "_WTF?!?_", you ask.

_"Oh, yeah, that's a known issue.  Entropy comes from a camera pointed at my pet hamster. When he's asleep it always outputs the same ID.  We only treat that as a single collision, though, because nobody cares about how *many* duplicates there are after the first one happens."_

Reply:
@broofa, I'm afraid you are confusing probability with expected value. They are different from each other and both are important to describe a random ID generator. And the probability of at least one collision is a legit probability measure that a lot of applications will seek for. For example, if such probability is low enough, some developers might write code that simply ignores collisions, betting that no collision would occur. For such code, consecutive collisions after the first one do not make a lot of difference because the first single collision will have already broken the code. This is presumably the way UUIDv4 is handled in many real world applications as of now.

Reply:
Great discussion!

I have always envisioned discussing the `counter` (a.k.a the old `clock sequence` verbiage I used pre-Draft 03) as part of the standard as it solves a very important problem that implementers will face with sortable time-based UUIDs.

---

In the [super old comparison documents from 16 months ago](https://github.com/uuid6/uuid6-ietf-draft/tree/master/research) I discuss some legacy and current non-standard implementations counter positions.

To summarize this the known approaches to the counter:
1. **Dedicated Counter Bits**
    - These usually are at the end of the layout. (e.g last X number of bits.) 
    The problem here is that most of the layouts that implement this end counter do not utilize random data. in other words `timestamp|staticdata|counter`. 
    When introducing random data as the next component after the timestamp in order to sort properly; one must freeze the random per timestamp tick and start incrementing the counter. Alternative is item 2 (see next numbered bullet.)
    - For others that do support random, which is what we want for this spec, sorting works better when we utilize the components of `timestamp|counter|random` and place the counter immediately after the timestamp.
   Instead of freezing the random one can generate new random for each counter tick within the same timestamp tick.
   This increases collision resistance, reduces guessability and sorts better in implementation testing. 
        - Then gets us to the question of: "How many bits are enough for millisecond level counting" I have seen anywhere from 8 to 24 with millisecond precision. For Draft 02, millisecond precision of UUIDv7 it was deemed that 12 bits was likely more than adequate.
    - Cons for both methods: Less bits for random, dedicated bits for something a system that only generates the one-off UUID do not ever utilize properly. Possible guidance on that problem is to specify for those that don't expect to need a counter can just use this as more random bits.
    - Other notes: A random initialize of the counter per timestamp tick increases probability of counter rollover; thus as a group in some threads here on GitHub we discussed that counters should start at 0 to maximize the space.
2. **Monotonic Random**: The alternative I have seen, as used by a few implementations, is to incorporate the counter into the random data. That is, if the timestamp has not incremented, freeze the random and then increment it by +1. Continue until the timestamp moves forward and then generate a new random. This is similar to the first numbered item on this list but reduces dedicated bits and replaces them with random. There is no rollover issue but this does increase guessability and as we discussed above. "How much collision resistance do we have when we monotonically increment the random?" The document provided by @fabiolimace is one such example that this could totally be fine.

---

Now this does only include a single node generating UUIDs. For the purpose of distributed nodes, of course the probability of collision increases as we add nodes. However, the solution to this problem was one of two options:
1. **Shared Knowledge System**: All nodes consult a central repository and ensure their generated value is unique before inserting into the DB. The problem with this is that very large systems can run into a [bottleneck](https://instagram-engineering.com/sharding-ids-at-instagram-1cf5a71e5a5c). As such in Draft 01 we adopted the logic that shared knowledge systems are not something we have in mind for the document.
2. **Machine ID**: If distributed nodes are an absolute, one should dedicated a portion of the random data to a non-descript machine ID. This ensures the bitspace of the machine generating UUIDs does not conflict with any other machine generating UUIDs. This is something I detailed in [Draft 02](https://datatracker.ietf.org/doc/html/draft-peabody-dispatch-new-uuid-format-02#section-7) and feel should be brought back into Draft 03.

---

In conclusion:
- I would like to include a writeup like this comment into the Draft 03 section, Monotonicity and Counters, which details both approaches to counters and what to do if an implementation does not want a counter.
   - My question for the group: is there any downside to letting an implementer choose any of these for their UUIDv7? i.e. no counter, explicit counter after timestamp (with some guidance on length), or monotonic random?
- Furthermore, I would like to modify Draft 03, Shared Knowledge, to include the text from Draft 02 on Distributed UUID Generation (possibly renaming the section to the draft 02 name as well)

Let me know what you think!

Reply:
Thank you @kyzer-davis for the great summary!

### Counters

As an implementer of a general-purpose UUID library, I would default to the monotonic random approach if this doesn't do a lot of harm to the universal collision resistance. A monotonic random field acts just as an entropy field when a user needs one or less UUIDv7 per millisecond, while it can also serve those who need multiple sortable UUIDs per millisecond. My feeling is that a monotonic counter doesn't sacrifice collision resistance, and the article @fabiolimace referred to exemplified this, but I think we need some mathematical grounds to conclude.

As for the counter rollover, if a monotonic random field doesn't sacrifice collision resistance, we might be able to spare as many bits as possible to the monotonic random counter to lower the probability of counter overflow. I am currently leaning towards this approach. In the rare circumstances where the counter does overflow, in my opinion a generator can just stall and wait for the next millisecond, but we have some discussion over this in #51.

Yes, the monotonic random field does damage to unguessability. So I think it's a good idea to reserve some tail bits as a randomness field to guarantee a certain level of unguessability at the spec level. Unguessability itself does not ensure security nor should be the first priority of UUIDv7, so a relatively small number of bits (say, 16 or 32 bits) will suffice in my opinion.

### Shared knowledge and machine IDs

I think it's better to eliminate these concepts from UUIDv7 and have applications that need these use UUIDv8 instead. Shared knowledge and machine IDs are very application-specific and have adverse impact on the uniqueness across applications, but such characteristics will not suffer UUIDv8 because UUIDv8 is totally application-specific by design. They will be good friends😃 This approach clarifies the roles of UUIDv7 and UUIDv8. We can focus on the universal uniqueness with UUIDv7, and UUIDv8 will get more adoptions by applications that are keen on the local uniqueness.

Reply:
@LiosK I'm continuing to ponder this.  I understand your point about how the characteristic of most interest to people is the likelihood of any sort of collision rather than the exact # of collisions.

But there's a fine line between "will there be collisions" and "how often will there be collisions".  As you scale the complexity of these use cases to include more nodes, with more ids, the "burstiness" of counter group collisions stops being a less of a factor, and the # of ids involved in such collisions starts will start to matter more.

Ultimately, it probably doesn't matter all that much.  The `rand_*` field should provide more than enough entropy to assuage people's concerns.  And if we take the ULID approach of treating that field as a randomly-seeded counter, then  I'm happy to continue using the birthday problem approach to collision probability calculations and ignore this issue altogether. :-)

Reply:
I didn't have time to hit this topic today but I will commit some text changes tomorrow during my working session that the group here can review. 

Good point @LiosK, I can add steering guidance towards UUIDv8 for the topic of shared knowledge and machine ID.

Reply:
BTW, I think we've missed @fabiolimace's point in [a separate comment](https://github.com/uuid6/uuid6-ietf-draft/pull/58#issuecomment-1046315626): extra timestamp precision in `rand_a`. Such additional precision might be particularly useful in some use cases (e.g. sorting IDs generated by multiple threads on a single machine). For this to work, the extra precision has to be placed right after `unix_ts_ms` and before the counter. So, it might be best for the spec to specify the order of fields: `unix_ts_ms` -> optional (variable-length) extra timestamp precision -> optional (variable-length) counter -> required? randomness. Any combination of these, placed in this order, will meet the objective of universally time-ordered UUID.

The sub-millisecond fraction representation might be useful for such an additional precision field because this encoding can guarantee the sortability of timestamps with a variety of size. The fractional encoding seems to have puzzled so many people and could be too difficult to require as a standard, but that shouldn't be as big an issue if the encoding is used only by an optional field.

P.S. I am very concerned about zero-starting counters because this is very difficult to operate safely. Accurate estimate of the number of IDs generated per millisecond is necessary to determine an appropriate size of the counter, which is not feasible in most cases. I tried to implement a dynamically sized counter whose size is adjusted when counter overflows are frequently observed, but this feature doubled the lines of code of a UUIDv7 generator. I think this doesn't pay off.

Reply:
@LiosK, I admit I was thinking generators on different machines, not different processes of a single machine. It guarantees some sort order on the same machine. Thank you! 

> the ULID approach of treating that field as a randomly-seeded counter

@broofa, this is exactly what a monotonic ULID is. It hits the bull's eye and is funny at the same time! hehe

Reply:
I am not going to resolve the thread but I did commit the discussed changes on the topic of counters and distributed node generation. Please take a look at the latest Draft 03 text post [`kydavis-2-23-2022 commit`](https://github.com/uuid6/uuid6-ietf-draft/pull/58/commits/600f55e28609f7cd1cbdeaa98dbec639ea087550) @broofa, @fabiolimace, and @LiosK

Reply:
1. I opened a discussion thread to continue this topic after this PR is closed. https://github.com/uuid6/uuid6-ietf-draft/issues/60
2. I put a few more layers into the monocity and counters section within this PR.